### ISSUE TITLE
V6 no way to use my own ODataPathRouteConstraint

### ISSUE THREAD

**Issue by rwmcdonald** [OData/WebApi#826]
In v6 it is impossible to provide my own ODataPathRouteConstraint.

I checked the [source](https://github.com/OData/WebApi/blob/OData60/OData/src/System.Web.OData/Extensions/HttpConfigurationExtensions.cs#L536) for MapODataServiceRoute and I see that it just uses the default:

`
            ODataPathRouteConstraint routeConstraint = new ODataPathRouteConstraint(routeName);
`

I expect this should be using GetService, or provide another way of setting my own customized ODataPathRouteConstraint.

Thanks,
Rob

**Comment by rwmcdonald** [comment]
BUMP

Just wondering if this is indeed a bug, or if I'm doing it wrong.

I have overridden the ODataPathRouteConstraint in much the same way as is done in the DynamicEdmModelCreation project here: https://github.com/OData/ODataSamples/blob/master/WebApi/v4/DynamicEdmModelCreation/DynamicEdmModelCreation/CustomODataPathRouteConstraint.cs

Is this no longer possible when using the DI methods provided in v6?

Thanks for your time,

**Comment by dmcallis** [comment]
Yes, I have the same problem. I think this is indeed a bug.

**Comment by oucfei** [comment]
Hi @rwmcdonald  just wondering were you able to solve this issue or have any work around? We also have our own ODataPathRouteConstraint in our project, and is planning to move to V6.

**Comment by rwmcdonald** [comment]
I was unable to use my own ODataPathRouteConstraint. However, I was able to workaround the issue.

It's been quite a long time, but I believe my ODataPathRouteConstraint was doing two things: Setting the model and setting the action/controller. 

I fixed it by creating a wrapper class to set the model and calling it per request as described by @VikingsFan  here: https://github.com/OData/WebApi/issues/816#issuecomment-249758634

And also built MyDynamicDataRoutingConvention implementing IODataRoutingConvention that selects the controller/action dynamically.

My final code looks something like the following:

`			IList<IODataRoutingConvention> routingConventions = ODataRoutingConventions.CreateDefault();
			routingConventions.Insert( 0, new MyDynamicDataRoutingConvention() );`
		
			
	var route = config.MapODataServiceRoute( "MyDynamic", "MyDynamic", containerBuilder =>
		containerBuilder.AddService<MyDynamicModelWrapper>( ServiceLifetime.Scoped )
		.AddService( ServiceLifetime.Scoped, sp => sp.GetService<MyDynamicModelWrapper>().GetModel )
		.AddService<ODataRoute>( ServiceLifetime.Scoped, sp => new ODataRoute( "MyDynamic", new ODataPathRouteConstraint( "MyDynamic" ) ) )
		.AddService<IEnumerable<IODataRoutingConvention>>( ServiceLifetime.Scoped, sp => routingConventions )
		.AddService<ODataUriResolver>( ServiceLifetime.Singleton, sp => new UnqualifiedCallAndEnumPrefixFreeResolver() )
	);

**Comment by oucfei** [comment]
@rwmcdonald thanks! I will try this out

**Comment by rbmanian75** [comment]
@rwmcdonald this one does not work for me.  My custom ODataRoute with ODataPathRouteConstraint is not injected.  But model is injected properly

**Comment by rmcdonald83** [comment]
Yes, as far as I know you cannot inject your own ODataPathRouteConstraint. I was able to workaround that by creating my own implementation of IODataRoutingConvention. You can use that to set the controller/action.

Would that work for you? If not, what does the ODataPathConstraint do that you require?

**Comment by rbmanian75** [comment]
I tried creating own IODataRoutingConvention. But to make it work we need custom odata route. The issue is explained here https://github.com/OData/ODataSamples/issues/50

**Comment by rbmanian75** [comment]
We can solve it by creating one more extension method as follows. By doing this we can add our own custom odataroute and odatapathconstraint. I dont much about github flow on how to create a pull request. 
`
/// <summary>
        /// Maps the custom OData route and the OData route attributes.
        /// </summary>
        /// <param name="configuration">The server configuration.</param>
        /// <param name="routeName">The name of the route to map.</param>
        /// <param name="routePrefix">The prefix to add to the OData route's path template.</param>
        /// <param name="configureAction">The configuring action to add the services to the root container.</param>
        /// <param name="route">The  <see cref="ODataRoute"/> </param>
        /// /// <returns>The added <see cref="ODataRoute"/>.</returns>
        public static ODataRoute MapODataServiceRoute(this HttpConfiguration configuration, string routeName,
            string routePrefix, Action<IContainerBuilder> configureAction, ODataRoute route)
        {
            if (configuration == null)
            {
                throw Error.ArgumentNull("configuration");
            }

            if (routeName == null)
            {
                throw Error.ArgumentNull("routeName");
            }

            if (route == null)
            {
                throw Error.ArgumentNull("route");
            }

            // 1) Build and configure the root container.
            IServiceProvider rootContainer = configuration.CreateODataRootContainer(routeName, configureAction);

            // 2) Resolve the path handler and set URI resolver to it.
            IODataPathHandler pathHandler = rootContainer.GetRequiredService<IODataPathHandler>();

            // if settings is not on local, use the global configuration settings.
            if (pathHandler != null && pathHandler.UrlKeyDelimiter == null)
            {
                ODataUrlKeyDelimiter urlKeyDelimiter = configuration.GetUrlKeyDelimiter();
                pathHandler.UrlKeyDelimiter = urlKeyDelimiter;
            }

            // 3) Resolve some required services and create the route constraint.
            ODataPathRouteConstraint routeConstraint = new ODataPathRouteConstraint(routeName);

            // Attribute routing must initialized before configuration.EnsureInitialized is called.
            rootContainer.GetServices<IODataRoutingConvention>();

            // 4) Resolve HTTP handler, create the OData route and register it.

            HttpRouteCollection routes = configuration.Routes;

            routes.Add(routeName, route);
            return route;
        }
`
